sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1304,10 +1307,27 @@ object CheckCartesianProducts extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  /**
+   * Check if a join contains PythonUDF in join condition.
+   */
+  def hasPythonUDFInJoinCondition(join: Join): Boolean = {
+    val conditions = join.condition.map(splitConjunctivePredicates).getOrElse(Nil)
+    conditions.exists(HandlePythonUDFInJoinCondition.hasPythonUDF)
+  }
+
   def apply(plan: LogicalPlan): LogicalPlan =
     if (SQLConf.get.crossJoinEnabled) {
       plan
     } else plan transform {
+      case j @ Join(_, _, _, _) if hasPythonUDFInJoinCondition(j) =>
```

Comment:
Move the check logic only in `CheckCartesianProducts ` as our discussion in https://github.com/apache/spark/pull/22326#discussion_r220236374 and add corresponding UT in CheckCartesianProductsSuite, the checking order is PythonUDF first and then the original checking for cartesian products.